### ISSUE TITLE
Allow "p" in Py_BuildValue

### ISSUE THREAD

**Issue by pablogsal** [python/cpython#89488]
BPO | [45325](https://bugs.python.org/issue45325)
--- | :---
Nosy | @vstinner, @serhiy-storchaka, @JelleZijlstra, @pablogsal, @godlygeek
PRs | <li>python/cpython#28634</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-09-29.18:08:04.356>
labels = []
title = 'Allow "p" in Py_BuildValue'
updated_at = <Date 2022-02-01.03:59:42.971>
user = 'https://github.com/pablogsal'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-01.03:59:42.971>
actor = 'JelleZijlstra'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2021-09-29.18:08:04.356>
creator = 'pablogsal'
dependencies = []
files = []
hgrepos = []
issue_num = 45325
keywords = ['patch']
message_count = 18.0
messages = ['402897', '402899', '402900', '402901', '402902', '402904', '402910', '402915', '402917', '402918', '402921', '402923', '402924', '402938', '402941', '402951', '402965', '412234']
nosy_count = 5.0
nosy_names = ['vstinner', 'serhiy.storchaka', 'JelleZijlstra', 'pablogsal', 'godlygeek']
pr_nums = ['28634']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue45325'
versions = []
```

</p></details>

### Linked PRs
* gh-135610
* gh-135843

**Comment by vstinner** [comment]
I raise the priority to release blocker since @serhiy-storchaka and @JelleZijlstra have concerns about this new feature.

@serhiy-storchaka wrote on the PR:

> This PR was merged, but the discussion on https://github.com/python/cpython/issues/89488 was not finished. I have concerns that this is a dangerous feature.

**Comment by godlygeek** [comment]
Those concerns continue to not make any sense to me. Don't they apply equally well to adding _any_ format code to `Py_BuildValue`? Or in fact to using `Py_BuildValue` at all?

Imagine that the `d` format code didn't exist, and we were suggesting adding it. Wouldn't all of the arguments that you're making for `p` being unsafe also apply to `d`? If someone passes an `int` or a `float` it would be memory unsafe, and instead of passing `x` using the `d` format specifier you could instead use the `N` format specifier and pass `PyFloat_FromDouble(x)`.

**Comment by pablogsal** [comment]
> I stumbled upon the PR and tend to agree with Serhiy that this could be a source of nasty bugs. Passing the wrong type to a va_arg() argument is undefined behavior (e.g. https://wiki.sei.cmu.edu/confluence/display/c/EXP47-C.+Do+not+call+va_arg+with+an+argument+of+the+incorrect+type), and it does seem easy to do that by accident for an int-like type. For printf() compilers implement custom checks that make sure the types are right, but we don't get that luxury with Py_BuildValue(). It's also easy to make a bool yourself with the O code and Py_True/Py_False, so I don't think the additional code is worth the risk.

Yeah I don't agree with this view. I can use this argument to whatever other parameter type to `Py_BuildValue` and it holds true so not sure why now this is different than the others. The only argument that I can think of is that "adding more potential problems is a problem" or that somehow this is a problem that will be **so much more** apparent than the rest but I would argue that none of these apply.

**Comment by pablogsal** [comment]
Also @serhiy-storchaka said:

> But I am not in strict opposition against this feature. ... Now you know about this, and you can decide whether the benefit is larger than the risk of potential damages. To me they look equally small.

So I haven't read this discussion as controversial to the point that we couldn't merge with two approvals (https://github.com/python/cpython/pull/28634) . @serhiy-storchaka even reviewed the PR

**Comment by serhiy-storchaka** [comment]
I don't want to block this feature. It has its value, it's clear and it can be convenient. I just wasn't sure that those who approved the PR were aware of the concerns. If they were aware, all is correct.

It might be worth adding a special note, but there are no notes for other formats, so it might be unnecessary. Let's see how many new errors using this feature will add, and if there are a lot, then we'll add a warning.

**Comment by godlygeek** [comment]
> It might be worth adding a special note, but there are no notes for other formats, so it might be unnecessary.

There aren't any notes for [any `Py_BuildValue` specifiers](https://docs.python.org/3/c-api/arg.html#c.Py_BuildValue), but there is a note for one of the [`PyArg_Parse`](https://docs.python.org/3/c-api/arg.html#c.PyArg_Parse) specifiers, so there's a precedent for it.

I do think it could be reasonable to add a small note, like:
```reST
   .. note::
      This format does not accept arbitrary boolean expressions, only ints.
      You can convert a pointer or a floating point number to an int suitable
      for this format with ``(x) ? 1 : 0`` or with ``!!x``.
```

**Comment by serhiy-storchaka** [comment]
Personally, I  do not think there is a need for a backtrack. We have more dangerous things in the C API. But I like a note proposed by @godlygeek.

**Comment by JelleZijlstra** [comment]
I looked at this a few years ago and I don't feel that strongly. I think what makes this format a bit riskier than other int formats is that relative to codes accepting numbers, it's easier to pass some random object (e.g., a pointer) and assume it will work like a boolean.

I'm OK with keeping it though. I think it would be good to add a warning to the Py_BuildValue docs that passing the wrong type may cause bad consequences, and sometimes it won't be caught by the compiler or the runtime and may cause UB.

**Comment by vstinner** [comment]
I'm in favor of having "p" format. But as I wrote previously, the concern can and should be addressed by the documentation.

```
      This format does not accept arbitrary boolean expressions, only ints.
      You can convert a pointer or a floating point number to an int suitable
      for this format with ``(x) ? 1 : 0`` or with ``!!x``.
```

`!!x` is a bit mysterious. Maybe `x != 0` and `ptr != NULL` are less surprising.

**Comment by pablogsal** [comment]
Ok I will do a PR to the documentation 👍

**Comment by hugovk** [comment]
(Marked as deferred-blocker, docs update doesn't need to block 3.14 beta 1.)

**Comment by encukou** [comment]
I've added the warning.
Thank you for the feature, Pablo!
